runtime/src/legacymm/cpp/Memory.cpp

```diff
@@ -2005,22 +2005,49 @@ MemoryState* initMemory() {
   memoryState->tlsMap = konanConstructInstance<KThreadLocalStorageMap>();
   memoryState->foreignRefManager = ForeignRefManager::create();
   bool firstMemoryState = atomicAdd(&aliveMemoryStatesCount, 1) == 1;
-  if (firstMemoryState) {
+  switch (Kotlin_getDestroyRuntimeMode()) {
+    case DESTROY_RUNTIME_LEGACY:
+      firstRuntime = firstMemoryState;
+      break;
+    case DESTROY_RUNTIME_ON_SHUTDOWN:
+      // Nothing to do.
+      break;
+  }
+  if (firstRuntime) {
 #if USE_CYCLIC_GC
     cyclicInit();
 #endif  // USE_CYCLIC_GC
     memoryState->isMainThread = true;
   }
+  if (!firstRuntime && firstMemoryState) {
```

Comment:
This part looks totally cryptic to me. Which problem does it solve?

Reply:
"Migrating" main thread.

Suppose we had a main thread, then this thread dies, and then later another thread appears and initializes Kotlin runtime on it.
In legacy mode this new thread becomes the main thread, and all the globals are freshly initialized on it.
I mimic this behaviour in the new mode: this new thread becomes the main thread, globals are not freshly initialized, but are now accessible on it. For the latter to work correctly I use `synchronize` in `deinitMemory` to publish all refcounts of globals and their object subgraphs, and `synchronize` here to acquire them.

If the main thread died while there were other threads, then in legacy mode globals become inaccessible (until all the threads die and a new one is born), and so it is in the new mode.

For experimental MM this behaviour is unneeded, because globals will be accessible from any thread regardless.

Reply:
Doesn't sound like worth fixing to me.

> Suppose we had a main thread, then this thread dies

Things are already pretty broken here. When K/N runtime considers some thread "main" while it is actually not a platform main thread, it is weird and in fact a bug. And platform main thread usually doesn't die.
Changing "main" thread during single execution is awkward too.

On the other hand, even if the solution is correct, it is still quite hard to reason about, considering all the memory model restrictions and GC implementation details. This just adds one more way for migrating thread-local objects to other threads.

Reply:
Ok, removed this logic.